haproxy_exporter.go

```diff
@@ -243,24 +247,65 @@ func (e *Exporter) Collect(ch chan<- prometheus.Metric) {
 	e.collectMetrics(ch)
 }
 
+func fetchHTTP(uri string, timeout time.Duration) func() (io.ReadCloser, error) {
+	client := http.Client{
+		Transport: &http.Transport{
+			Dial: func(netw, addr string) (net.Conn, error) {
+				c, err := net.DialTimeout(netw, addr, timeout)
+				if err != nil {
+					return nil, err
+				}
+				if err := c.SetDeadline(time.Now().Add(timeout)); err != nil {
+					return nil, err
+				}
+				return c, nil
+			},
+		},
+	}
+
+	return func() (io.ReadCloser, error) {
+		resp, err := client.Get(uri)
+		if err != nil {
+			return nil, err
+		}
+		if !(resp.StatusCode >= 200 && resp.StatusCode < 300) {
+			return nil, fmt.Errorf("HTTP error %d", resp.StatusCode)
+		}
+		return resp.Body, nil
+	}
+}
+
+func fetchUnix(u *url.URL) func() (io.ReadCloser, error) {
+	return func() (io.ReadCloser, error) {
+		f, err := net.Dial("unix", u.Path)
+		if err != nil {
+			return nil, err
+		}
+		cmd := "show stat\n"
+		n, err := io.WriteString(f, cmd)
+		if err != nil {
+			return nil, err
```

Comment:
Wouldn't it be necessary to close the connection here given an error is returned? Same below.